[Guidelines] Add writing examples page

miukimiu opened the pull request:
### Summary

This PR adds:

- A new **Examples** page into Guidelines -> Writing
- Improves the `<GuideRuleExample />` component
- Adds a new method to create tabbed pages: `createTabbedPage()`

### New examples pages

This PR adds a new examples page into Guidelines. The examples show some EUI components with writing examples. All the examples were provided by @gchaps and @lcawl.

I just improved the design and the usage of some components.

![Writing-Elastic-UI-Framework (3)](https://user-images.githubusercontent.com/2750668/142210674-a4d8071a-9f5f-489e-b521-4dce9a2cffe2.png)


### Improved `GuideRuleExample` component

Removed the left and right padding for the text:

<img width="605" alt="left-right-padding@2x" src="https://user-images.githubusercontent.com/2750668/142209416-a450fdaa-7b4a-49ae-83c1-38a3babf785e.png">

Added a new type `<GuideRuleExample type="default" />`. It's also the default prop, so in case a  `type` is not passed the component gets the default state. Now we're not stuck to **"Do"** and **"Don't"**.


<img width="1250" alt="Screenshot 2021-11-17 at 13 26 40" src="https://user-images.githubusercontent.com/2750668/142209740-da05b226-2b64-4d45-8742-a23b971156e2.png">


### New method to create tabbed pages.

This PR adds a new method `createTabbedPage()` in `routes.js`. This method allows the creation of a tabbed page. 

```js
const navigation = [
  {
    name: 'Guidelines',
    items: [
      createExample(AccessibilityGuidelines, 'Accessibility'),
      {
        name: 'Colors',
        component: ColorGuidelines,
      },
     // here's the method
      createTabbedPage('Writing', [
        {
          title: 'Guidelines',
          page: WritingGuidelines,
        },
        { title: 'Examples', page: WritingExamples },
      ]),
    ],
  },
...
];
```

This method creates a tabbed page and the side nav gets 3 levels.

![Writing-Elastic-UI-Framework (1)](https://user-images.githubusercontent.com/2750668/114913327-fd3d7880-9e18-11eb-820c-076886f4b59a.png)


### Checklist

- [x] Check against **all themes** for compatibility in both light and dark modes
- ~[ ] Checked in **mobile**~
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
- ~[ ] Props have proper **autodocs** and **[playground toggles]~(https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#adding-playground-toggles)**
- [x] Added **[documentation](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md)**
- ~[ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for the any docs examples~
- ~[ ] Added or updated **[jest tests](https://github.com/elastic/eui/blob/master/wiki/testing.md)**~
- ~[ ] Checked for **breaking changes** and labeled appropriately~
- ~[ ] Checked for **accessibility** including keyboard-only and screenreader modes~
- ~[ ] A **[changelog](https://github.com/elastic/eui/blob/master/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately~

miukimiu commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

cchaos commented:
OGM, I actually need this functionality (the new generic tabbed guideline pages). I'm working on re-writing our Getting Started & Theming docs and really want to put in tabs that show up in the docs. Can we maybe make this PR just about creating the tabs for docs?

miukimiu commented:
Yes, I'll open a PR with the functionality.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

thompsongl commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

gchaps commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+            <EuiModalFooter>
+              <EuiButtonEmpty>Continue editing</EuiButtonEmpty>
+              <EuiButtonEmpty>Discard changes</EuiButtonEmpty>
+              <EuiButton fill>Keep changes </EuiButton>
```
Can we make this button longer.  It cuts off in the image.

gchaps commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+          type="do"
+          text="Prefer this format for a required field."
+        >
+          <EuiFormRow label="Name" helpText={<span>Name is required.</span>}>
```
As this is a validation message, this should be in red.

gchaps commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+            label="Name"
+            helpText={
+              <span>
+                Choose a different address as this one already exists.
```
This is also a validation message.

kellyemurphy commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+            <EuiFieldText />
+          </EuiFormRow>
+        </GuideRuleExample>
+        <GuideRuleExample
+          panelDisplay="block"
+          type="dont"
+          text="Use generic labels."
+        >
+          <EuiFormRow>
+            <EuiCheckbox
+              onChange={() => {}}
+              id={htmlIdGenerator()()}
+              label="Combine values in other bucket"
+            />
+          </EuiFormRow>
+          <EuiFormRow label="Bucket label">
```
```suggestion
        <GuideRuleExample
          panelDisplay="block"
          type="do"
          text="Use specific labels."
        >
          <EuiFormRow>
            <EuiCheckbox
              onChange={() => {}}
              id={htmlIdGenerator()()}
              label="Limit incoming traffic"
            />
          </EuiFormRow>
          <EuiFormRow 
                      label="Traffic source"
                      helpText={
              <span>Enter CIDR or IP address, for example 192.168.132.6/22.</span>
            }>
```

I hope I've done this right. I'd prefer to get rid of the checkbox for this entirely, but I wasn't sure If I could, so I just changed the wording.

constancecchen commented:
Searching and clicking into 3rd level nav items should be working now!

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

lcawl commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+        <GuideRuleExample
+          panelDisplay="block"
+          type="do"
+          text='Start with a keyword instead of "If true" or "If enabled".'
```
In this case, I'm not sure which "keyword" is meant in the "use recommended defaults" switch. Or was the rollover text supposed to be part of the switch?
```suggestion
          text='Start with a keyword instead of "If true" or "If enabled".'
```

lcawl commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+            <EuiModalBody>
+              <EuiText>
+                <p>
+                  You can keep or discard your changes on return to view mode.
```
It might just be me, but I don't understand "on return to view mode". For the purposes of a simple example, can we just simplify this statement? Maybe something like: 
```suggestion
                  When you exit editing mode, you can keep or discard your changes.
```

lcawl commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+          </EuiPanel>
+        </GuideRuleExample>
+
+        <GuideRuleExample type="do" text="Some cases requrie three buttons.">
```
```suggestion
        <GuideRuleExample type="do" text="Some cases require three buttons.">
```

lcawl commented on src-docs/src/views/guidelines/writing_examples.js:
```diff
@@ -0,0 +1,730 @@
+        </GuideRuleExample>
+      </GuideRule>
+
+      <GuideRule description="Fields are more usable when you omit placeholder text.">
```
>      <GuideRule description="Fields are more usable when you omit placeholder text.">

I think this example can be omitted since it's a repeat of an earlier example.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

miukimiu commented:
Thanks, @constancecchen for fixing the searching and clicking into 3rd level nav items! 🎉

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

miukimiu commented:
@gchaps and @lcawl can you take a look at all the design changes [here](https://eui.elastic.co/pr_4720/#/guidelines/writing/examples) and let me know if there's something you want to change? 

I created a new example type called "default."  This way, we don't need to use the "Do" or "Don't"  and have a green background. This change was suggested by @cchaos, and I think it works better to showcase examples.

![default](https://user-images.githubusercontent.com/2750668/142252335-06886576-bbd3-48da-96bf-5a0c993413ca.png)

I also changed the icons for the empty prompts examples and made sure all the examples have the same height when they're positioned side by side.

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

lcawl reviewed (approved):
LGTM

gchaps reviewed (approved):
thanks @miukimiu for all your work on this.  I really like that it's not do and don'ts. LGTM

cchaos commented on src-docs/src/components/guide_page/guide_page_chrome.js:
```diff
@@ -192,6 +202,22 @@ export class GuidePageChrome extends Component {
     return sideNavSections;
   };
 
+  searchSubSections = (searchTerm, navItem) => {
```
I don't think this is allowing subsequent sub-sections to show matching search terms. For instance, When I search for `switch`, Only the `Writing / Examples / Switches` item is showing, but we should also see `Forms / Form controls / Switch`
<img width="275" alt="Screen Shot 2021-11-18 at 10 46 32 AM" src="https://user-images.githubusercontent.com/549577/142456434-047a74a5-0568-4a11-b969-8c092c8472af.png">

cchaos replied:
One more side effect I noticed is that the `NEW` badge is now on the left instead of the right. Was this intentional?
<img width="242" alt="Screen Shot 2021-11-18 at 11 43 52 AM" src="https://user-images.githubusercontent.com/549577/142458808-d85eca4e-21d7-4ee7-b1e0-3709c3f611d0.png">

cchaos replied:
Should we exclude the nav items that come from the tabs from search? I'm thinking long-term when we add these for all the "Guidelines vs Examples" tabs, may be overwhelming.
<img width="241" alt="Screen Shot 2021-11-18 at 11 44 57 AM" src="https://user-images.githubusercontent.com/549577/142459118-e598d6f2-7d15-4b34-9a15-eff8ad24aabd.png">

cchaos commented on src-docs/src/components/guide_tabbed_page/guide_tabbed_page.tsx:
```diff
@@ -0,0 +1,92 @@
+import React, { FunctionComponent } from 'react';
+import { Switch, Route, withRouter, Redirect } from 'react-router-dom';
```
I tried adding the `<Link>` from `react-router-dom` in the new Writing Examples page, but it complained that it couldn't be used outside of `Router`. Can we fix that?

cchaos reviewed (changes requested):
I've pushed you a commit that has some design changes. Mainly just cleaning up some sizing and such and adding links to the actual components. I think you can also delete the old `writing.js` file now.

The rest of the comments are around the new nav functionality.

gchaps commented:
On this image, text is truncated on one of the buttons?  Can we made the modal wider.  Or should we come up with another example?

<img width="521" alt="Screen Shot 2021-11-18 at 9 18 45 AM" src="https://user-images.githubusercontent.com/33642766/142464675-4f932d18-3c58-41d9-a0f9-51cd5ea65c29.png">

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

cchaos commented:
@gchaps I already pushed a fix for that by allowing them to `wrap`. Still not _ideal_, but at least the text doesn't get cut off.
<img width="515" alt="Screen Shot 2021-11-18 at 12 32 01 PM" src="https://user-images.githubusercontent.com/549577/142466792-350978df-2481-46e0-9173-8c265ae5b7ae.png">

miukimiu commented:
Thanks @cchaos, I'm taking a look at the nav issues and deleting the `writing.js` file. 👍🏽

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

miukimiu commented:
Thanks, @cchaos. I went through your comments. 

> Also, can we remove the highlighting of the hashed navigation item?

I removed the highlighting.

> One more side effect I noticed is that the NEW badge is now on the left instead of the right. Was this intentional?

It was not intentional. It is now fixed! 😄 

> Should we exclude the nav items that come from the tabs from search?

I removed them, but I think we should rethink if we should allow or do not allow searching third levels. In your PR #5378 I think it would make sense to get search results for brand colors, shaded, and so on, when the colors nav have 3 levels (colors -> values -> brand colors). 

But we can rethink this when we support a new layout and move the search to the top `EuiHeader`.

> I tried adding the `<Link>` from react-router-dom in the new Writing Examples page, but it complained that it couldn't be used outside of Router. Can we fix that?

Fixed! It seems that getting the `h2` IDs to create the hashed navigations from the rendered components is not supported. If the component has a `themeContext` or  `routes` it fails. I was rendering it in `route.js` to get with regex all the IDs from an `h2`. But when it renders it fails because there's no context or routes yet... 

So the solution is to manually create the nav items by passing an array of objects with titles and ids for each section. 

I also improved the `GuideTabbedPage ` and `createTabbedPage` to support the layout you need for the colors section in #5378. I tested using the **Borders** page as a third tab:

![Writing-Elastic-UI-Framework (3)](https://user-images.githubusercontent.com/2750668/142632717-3d41f2cd-2f8a-4479-958b-10531fe78b49.png)

To create this example we would do the following in `routes.js`. Well... This can be improved. But I would continue this in your PR. The logic is there but it might require some adjustments. 

```js
const writingSection = createTabbedPage({
    title: 'Writing',
    pages: [
      {
        title: 'Guidelines',
        page: WritingGuidelines,
        sections: writingGuidelinesSections,
      },
      {
        title: 'Examples',
        page: WritingExamples,
        sections: writingExamplesSections,
      },
      {
        title: 'Borders',
        page: BordersInTab,
        sections: bordersSections,
      },
    ],
    isNew: true,
    description:
      'The border tokens contain both individual border property values and full shorthand border properties.',
    showThemeLanguageToggle: true,
    notice: <ThemeNotice />,
  });
```

Let me know if you find another issue.

cchaos commented:
Thank you for those updates to the new nav. I think there was some confusion on this part though:

>> Should we exclude the nav items that come from the tabs from search? I'm thinking long-term when we add these for all the "Guidelines vs Examples" tabs, may be overwhelming.

> I removed them, but I think we should rethink if we should allow or do not allow searching third levels.

I didn't mean to exclude the actual named sections but the names of the tabs themselves. The `Guidelines` and `Examples` words. If someone started typing `guide` they will eventually have a very long list of "Guideline" tabs that aren't very meaningful.

![Image 2021-11-20 at 10 31 04 AM](https://user-images.githubusercontent.com/549577/142732184-92476181-97dd-4537-ab78-41c5f7758c3b.jpg)

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_4720/

miukimiu commented:
@cchaos I just added the searching to 3rd levels sub-nav items back. 

> I don't think this is allowing subsequent sub-sections to show matching search terms. For instance, When I search for switch, Only the Writing / Examples / Switches item is showing, but we should also see Forms / Form controls / Switch

This is fixed. 


<img width="170" alt="Screenshot 2021-11-22 at 15 44 18" src="https://user-images.githubusercontent.com/2750668/142891848-7a7f6aff-3b6e-47a2-89d5-c11020d07faa.png">

> I didn't mean to exclude the actual named sections but the names of the tabs themselves. The `Guidelines` and `Examples` words. 

I couldn't solve this one. @constancecchen can you help me solve this?

cchaos reviewed (approved):
Thanks for making those fixes. Leaving out the tab names would be ideal, but I'm ok if we just do a follow up since this is the only page with that pattern right now.

Great job on these guidelines! And thanks for all the help writers!

miukimiu commented:
Thanks for the review @cchaos.

Thanks @constancecchen, @gchaps, and  @lcawl for all the teamwork! 🎉
